File: .tmp_sc4013/java_22/3980_minimum_operations_to_transform_binary_string.java
// CONFIG class=Solution method=minOperations types=None
// LeetCode 3980 - Minimum Operations to Transform Binary String
// https://leetcode.com/problems/minimum-operations-to-transform-binary-string/

class Solution {
    public int minOperations(String s1, String s2) {
        final int infinity = 1000000000;
        int[] dp = new int[] { 0, infinity };
        int n = s1.length();
        for (int i = 0; i < n; i++) {
            int[] next = new int[] { infinity, infinity };
            for (int forcedZero = 0; forcedZero <= 1; forcedZero++) {
                if (dp[forcedZero] == infinity) continue;
                char current = s1.charAt(i);
                if (forcedZero == 1) current = '0';
                int direct = dp[forcedZero];
                if (current == '0' && s2.charAt(i) == '1') direct++;
                else if (current == '1' && s2.charAt(i) == '0') direct = infinity;
                next[0] = Math.min(next[0], direct);
                if (i + 1 < n) {
                    int cost = dp[forcedZero] + 1;
                    if (current == '0') cost++;
                    if (s1.charAt(i + 1) == '0') cost++;
                    if (s2.charAt(i) == '1') cost++;
                    next[1] = Math.min(next[1], cost);
                }
            }
            dp = next;
        }
        return dp[0] == infinity ? -1 : dp[0];
    }
}

File: .tmp_rs4013/rust_check22/3980_minimum_operations_to_transform_binary_string.rs
struct Solution;
// LeetCode 3980 - Minimum Operations to Transform Binary String
// https://leetcode.com/problems/minimum-operations-to-transform-binary-string/

impl Solution {
    pub fn min_operations(s1: String, s2: String) -> i32 {
        const INFINITY: i32 = 1_000_000_000;
        let s1 = s1.into_bytes();
        let s2 = s2.into_bytes();
        let n = s1.len();
        let mut dp = [0, INFINITY];
        for i in 0..n {
            let mut next = [INFINITY, INFINITY];
            for forced_zero in 0..=1 {
                if dp[forced_zero] == INFINITY {
                    continue;
                }
                let mut current = s1[i];
                if forced_zero == 1 {
                    current = b'0';
                }
                let mut direct = dp[forced_zero];
                if current == b'0' && s2[i] == b'1' {
                    direct += 1;
                } else if current == b'1' && s2[i] == b'0' {
                    direct = INFINITY;
                }
                next[0] = next[0].min(direct);
                if i + 1 < n {
                    let mut cost = dp[forced_zero] + 1;
                    if current == b'0' {
                        cost += 1;
                    }
                    if s1[i + 1] == b'0' {
                        cost += 1;
                    }
                    if s2[i] == b'1' {
                        cost += 1;
                    }
                    next[1] = next[1].min(cost);
                }
            }
            dp = next;
        }
        if dp[0] == INFINITY {
            -1
        } else {
            dp[0]
        }
    }
}

fn main() {}
